### PR TITLE
Extract configs into separate Dataclass and simplify training pipeline 

### PR DESCRIPTION
**Pull request by workofart**
## Description
- Add config schema dataclass for `GenericTrainingConfig`, `TransformerTrainingConfig` and `CustomBpeConfig`
- Greatly simplify the trainer class (e.g. args, state management, 
- Make learning rate scheduler configurable by instantiating in the trainer class using kwargs from the configs
- Move the load_wiki_simple() and load_shakespeare_mini() data fetching functions into the `autograd/text/utils.py` to make the examples slim
- Refactor the forward function for Transformer-based models to implement the `AbstractLLMForwardFn` interface with a clean `train` and `sample` mode  interface

## Tests
Updated unit tests for all updated functionalities, especially the trainer class <-> config integration, to solidify the contract between them.

All examples training successfully.

### CODE REVIEW

File: autograd/nn.py [hunk 1]
```diff
@@ -997,15 +997,11 @@ class AbstractLLMForwardFn(ABC):
     """
 
     @abstractmethod
-    def sample(
-        self, model: Any, batch_data: Any, mode: str = "train"
-    ) -> Tuple[Any, Any]:
+    def sample(self, model: Any, batch_data: Any) -> Tuple[Any, Any]:
```
Comment (workofart) [inline autograd/nn.py:1000]: Unnecessary `mode` argument, it's only necessary in the `__call__` function below.

File: autograd/text/utils.py [hunk 2]
```diff
@@ -282,7 +283,7 @@ def sample_next_token(logits: np.ndarray, temp: float, k: Optional[int]) -> int:
     else:
         start_tokens = start_tokens or "<SOS>"
         output_ids = list(bpe.encode(start_tokens))
-        num_steps = max_length
+        num_steps = max_length - len(output_ids)
```
Comment (workofart) [inline autograd/text/utils.py:286]: Bug fix where the start_tokens length wasn't considered when we perform the inference, and the inference steps could have exceeded the max_seq_len of the model.

File: autograd/tools/trainer.py [hunk 3]
```diff
@@ -336,68 +410,3 @@ def grad_l2_norm(parameters: Dict[str, Tensor]) -> float:
         if p.grad is not None:
             grad_norm += (p.grad.data**2).sum()
     return grad_norm**0.5
-
-
-def load_model_and_optimizer(
```
Comment (workofart) [inline autograd/tools/trainer.py:341]: moved to inside the trainer class for more streamlined process and the user doesn't need to know about this.
